Added bulk card regeneration action in Team admin

kingraphaii opened the pull request:
> Added confirmation template (WIP)

Related issues: N/A

## Description:

This new feature allows admins to regenerate cards for Teams.

In the Django admin, users can select the teams to do the regen. for, then choose 'Bulk regenerate cards for members' in the *Actions* dropdown.

Task runs in background, so if working on large datasets, changes may take time to reflect.

## Screenshots/videos
![image](https://github.com/Umuzi-org/Tilde/assets/57295227/1285bfb8-4d83-437f-a186-4163cc97751c)

Users will need to confirm this action as it deletes any existing assignments to other cards:
![image](https://github.com/Umuzi-org/Tilde/assets/57295227/a7633d9d-0868-4f5f-bd0c-adb04b6e7bbd)


## I solemnly swear that:

- [X] My code follows the style guidelines of this project
- [X] I have merged the develop branch into my branch and fixed any merge conflicts
- [X] I have performed a self-review of my own code
- [X] I have commented my code in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have tested new or existing tests and made sure that they pass

kingraphaii committed:
Added bulk card regeneration action in Team admin

> Added confirmation template (WIP)

sheenarbw commented on backend/core/admin.py:
```diff
@@ -58,9 +59,29 @@ def deactivate_team_members(self, request, queryset: object):
+    def bulk_regenerate_cards_for_members(self, request, queryset: object):
+        if request.POST.get('post', None):
+            for team in queryset:
+                bulk_regenerate_cards_for_team(team)
```
This would be a very long running request. It should be handled by the dramatiq worker. Take a look at how the current delete and recreate cards api works. 

I would append one task for each user who needs to be regenerated.

sheenarbw replied:
So something like:

```
from long_running_request_actors import (
                delete_and_recreate_user_cards as actor,
            )

for user_id in active_memer_ids:
    actor.send_with_options(kwargs={"user_id": user.id})

```

kingraphaii replied:
What would you recommend, spawning `delete_and_recreate_user_cards` actor for each user in a team in a new actor, say `bulk_regenerate_cards_for_members`, or running that separate actor, doing all the regeneration for all users at once

kingraphaii replied:
Something like:

```python
@actor(max_retries=3)
def bulk_regenerate_cards_for_team(team):
    for user in team.users.filter(active=True):
        delete_and_recreate_user_cards.send_with_options(kwargs={"user_id": user.id})
```

sheenarbw replied:
Better to do it the way I suggested. It would be faster because regens can happen in parallel. Also if there are any issues with regenning one user's board then that wont totally derail the process for all subsequent users.

sheenarbw commented on backend/curriculum_tracking/card_generation_helpers.py:
```diff
@@ -266,7 +266,14 @@ def update_cards_accorrding_to_progress(user):
 
 def generate_all_content_cards_for_team(team):
-    todo
+    # todo
```
This function was meant to fail loudly instead of fail silently.
I don't think it's being used anywhere though.

If you don't like the `todo` you can explicitly raise a notimplemented exception. It amounts to the same thing. `todo` is just faster to type.

sheenarbw reviewed (changes requested):
Largely on the right track :) You're missing one important detail.

I also don't work with the admin panel enough to visualize the confirmation form you added there without rinning the code, but I'm sure it's fine.

sheenarbw commented on backend/curriculum_tracking/card_generation_helpers.py:
```diff
@@ -266,7 +266,14 @@ def update_cards_accorrding_to_progress(user):
+    pass
+
+
+def bulk_regenerate_cards_for_team(team):
```
Then this wont be needed any more

kingraphaii committed:
Moved regeneration to dramatiq worker
- new long running worker to regenerate cards
- added message on regen. action
- clean up regen. confirmation template

kingraphaii committed:
Added test for team admin actions

- needed as actions are now lazily loaded and it won't fail fast.

kingraphaii committed:
Misc: Remove redundant context var in confirmation template

- +Reset erroneous update to package-lock.json

kingraphaii committed:
Misc: Removed unused var assignment

Ngoakor12 reviewed (changes requested):
- Maybe add a proper title for action confirmation page(before "|")
![image](https://github.com/Umuzi-org/Tilde/assets/54069197/08a86d0d-565b-40a6-a8ab-8dc27dc7c736)


Otherwise I don't see anything bad

kingraphaii committed:
Fix:Confirmation page title

+ minor rewording of confirmation message

kingraphaii committed:
ignore .vscode dir

sheenarbw committed:
Merge branch 'develop' into bulk_regenerate_boards

sheenarbw commented on backend/core/tests/test_model_admin_actions_exist.py:
```diff
@@ -0,0 +1,12 @@
+
+
+class TestModelAdminActionsExist(TestCase):
+     def setUp(self) -> None:
```
If you don't make a setUp method then the super class's setup will get called by default. Did you mean to do something here?  This code doesn't do anything.
